orttraining/orttraining/test/distributed/pipeline_schedule_test.cc

```diff
@@ -0,0 +1,125 @@
+      EXPECT_EQ(forward_record_before_send, baseline_events[2 * s + 0][4 * b + 2]) << " batch " << b << " stage " << s;
+      EXPECT_EQ(forward_record_after_send, baseline_events[2 * s + 0][4 * b + 3]) << " batch " << b << " stage " << s;
+
+      EXPECT_EQ(backward_wait_before_recv, baseline_events[2 * s + 1][4 * b + 0]) << " batch " << b << " stage " << s;
+      EXPECT_EQ(backward_wait_after_recv, baseline_events[2 * s + 1][4 * b + 1]) << " batch " << b << " stage " << s;
+      EXPECT_EQ(backward_record_before_send, baseline_events[2 * s + 1][4 * b + 2]) << " batch " << b << " stage " << s;
+      EXPECT_EQ(backward_record_after_send, baseline_events[2 * s + 1][4 * b + 3]) << " batch " << b << " stage " << s;
+    }
+  }
+}
+
+TEST(Pipeline, ScheduleB8S3) {
+  const int num_batches = 8;
+  const int num_stages = 3;
+
+  // The event baselines at different stages are the same.
+  // The first 4 events are for the first computation on that stage.
+  // Similarly, the last 4 events are for the last computation on that stage.
+  // Below, we add comments to indicate which computation the events associated with.
+
+  // Format per line below:
+  //   waited event before Recv, waited event after Recv, recorded event before Send, recorded event after Send.
+  // The value "-1" means a NULL event; RecordEvent and WaitEvent do nothing for NULL events.
+  // Note that the computation pattern is
+  //   WaitEvent -> Recv -> WaitEvent -> FW/BW -> RecordEvent -> Send -> RecordEvent.
+  std::vector<int> forward_baseline_events_stage0{
+      -1, -1, 0, 1,    // FW00 @ stage 0
+      0, 1, 2, 3,      // FW01
+      2, 3, 4, 5,      // FW02
+      6, 7, 8, 9,      // FW03
+      10, 11, 12, 13,  // FW04
+      14, 15, 16, 17,  // FW05
+      18, 19, 20, 21,  // FW06
+      22, 23, 24, 25,  // FW07
+  };
+
+  std::vector<int> backward_baseline_events_stage0{
+      4, 5, 6, 7,      // BW00
+      8, 9, 10, 11,    // BW01
+      12, 13, 14, 15,  // BW02
+      16, 17, 18, 19,  // BW03
+      20, 21, 22, 23,  // BW04
+      24, 25, 26, 27,  // BW05
+      26, 27, 28, 29,  // BW06
+      28, 29, 30, 31   // BW07
+  };
+
+  std::vector<int> forward_baseline_events_stage1{
+      -1, -1, 0, 1,    // FW00 @ stage 1
+      0, 1, 2, 3,      // FW01
+      2, 3, 4, 5,      // FW02
+      8, 9, 10, 11,    // FW03
+      12, 13, 14, 15,  // FW04
+      16, 17, 18, 19,  // FW05
+      20, 21, 22, 23,  // FW06
+      24, 25, 26, 27,  // FW07
+  };
+
+  std::vector<int> backward_baseline_events_stage1{
+      4, 5, 6, 7,      // BW00
+      6, 7, 8, 9,      // BW01
+      10, 11, 12, 13,  // BW02
+      14, 15, 16, 17,  // BW03
+      18, 19, 20, 21,  // BW04
+      22, 23, 24, 25,  // BW05
+      26, 27, 28, 29,  // BW06
+      28, 29, 30, 31   // BW07
+  };
+
+  std::vector<int> forward_baseline_events_stage2{
+      -1, -1, 0, 1,    // FW00 @ stage 2
+      2, 3, 4, 5,      // FW01
```

Comment:
>2, 3 [](start = 6, length = 4)

Does the event distinguish fw pass or bw pass? I guess I am not clear if this '2' comes from the forward pass in stage 1 (FW01) or the backward pass in stage 2 (BW00)

Reply:
the variable name seems to indicate it's FW. event id itself is just an int and does not carry FW/BW semantics I think.

---
In reply to: [456110213](https://github.com/microsoft/onnxruntime/pull/4524#discussion_r456110213) [](ancestors = 456110213)

Reply:
Right. This Wait '2' is in the FW pass and it's waiting a Record '2' in other place to sent it the signal. But the Record '2' can be from the FW pass in stage 1 (FW01) or the BW pass in stage 2(BW00). So not sure which signal it is listening to. 

---
In reply to: [456111879](https://github.com/microsoft/onnxruntime/pull/4524#discussion_r456111879) [](ancestors = 456111879,456110213)

Reply:
Yes, event ID doesn't define forward or backward. It's just a format of mine; I choose to split events based on forward/backward `after events have been generated`.

`Wait 2` is on stage 2, which will be triggered by `Record 2` on stage 2 (in the first row of `backward_baseline_events_stage2`).


First of all, different stages are run by different processes, so their events are independent event they have the same ID. That is, Event 0 on stage 0 has nothing to do with Event 0 on stage 1. The format of these 4 ints is
```
  // Format per line below:
  //   waited event before Recv, waited event after Recv, recorded event before Send, recorded event after Send.
```
It means `2` is the event waited before Recv in FW01 on stage 2. Similarly, `3`, `4`, and `5` sequentially maps to `waited event after Recv`, `recorded event before Send`, and `recorded event after Send`.

---
In reply to: [456124585](https://github.com/microsoft/onnxruntime/pull/4524#discussion_r456124585) [](ancestors = 456124585,456111879,456110213)